app/src/main/java/cloudcity/Iperf3Monitor.java

```diff
@@ -310,10 +315,13 @@ public void propertyChange(PropertyChangeEvent evt) {
                                     stopParsingThread();
                                     shouldStop.compareAndSet(false, true);
 
+                                    long endTimestamp = System.currentTimeMillis();
                                     // Instantiate the POJO stuff holder
                                     MetricsPOJO values = new MetricsPOJO(
                                             new MetricsPOJO.DownloadMetrics(DLmin, DLmedian, DLmean, DLmax, DLlast),
-                                            new MetricsPOJO.UploadMetrics(ULmin, ULmedian, ULmean, ULmax, ULlast)
+                                            new MetricsPOJO.UploadMetrics(ULmin, ULmedian, ULmean, ULmax, ULlast),
+                                            testStartTimestamp,
+                                            endTimestamp
```

Comment:
Look at the whole piece.
```
                                    // Instantiate the POJO stuff holder
                                    MetricsPOJO values = new MetricsPOJO(
                                            new MetricsPOJO.DownloadMetrics(DLmin, DLmedian, DLmean, DLmax, DLlast),
                                            new MetricsPOJO.UploadMetrics(ULmin, ULmedian, ULmean, ULmax, ULlast),
                                            testStartTimestamp,
                                            endTimestamp
                                    );

                                    // Notify the completion listener
                                    if (completionListener != null) {
                                        completionListener.onIperf3TestCompleted(values);
                                    }

                                    // And set the new marker as 'no longer running'
                                    iperf3TestRunning.compareAndSet(true, false);
                                    // Reset the test start timestamp
                                    testStartTimestamp = 0;
```

While it is true that a test could start between a call to this `iperf3TestRunning.compareAndSet(true, false);` and resetting the `testStartTimestamp`, this AtomicBoolean **guarantees** that no other test can start while it's set.

Do you remember pestering me about removing redundant synchronization techniques and how they could all have been replaced by one? Well, the AtomicBoolean that was put into place was put in *exactly* to solve these kind of "what if..." worries while a test is running.

So no, a test most certainly won't start between capturing endTimestamp and resetting testStartTimestamp. 
In fact, to doesn't matter what happens with testStartTimestamp after we've captured the start/end pair. since the test timestamp is only valid until the next endTimestamp comes along.